File: Navigation/nav2_option/g1_sdk_bridge_cpp/tests/test_sdk_bridge.cpp
// SdkBridgeProcess の統合テスト。実スレッド + 実Unix domain socketで検証する。
//
// タイミング系のテストはsleepではなくWaitUntil()でポーリングし、CI環境差による
// flakyさを抑える。ただしスレッド+実ソケットを使うため、他のテストより本質的に
// 時間がかかる(数百ms程度)。

#include "g1_sdk_bridge/sdk_bridge_process.hpp"

#include <gtest/gtest.h>

#include <unistd.h>

#include <algorithm>
#include <chrono>
#include <cstdio>
#include <functional>
#include <memory>
#include <stdexcept>
#include <string>
#include <thread>
#include <vector>

#include "g1_sdk_bridge/ipc_transport.hpp"
#include "g1_sdk_bridge/protocol.hpp"

using namespace g1_sdk_bridge;
using namespace std::chrono_literals;

namespace {

void WaitUntil(const std::function<bool()>& predicate, std::chrono::duration<double> timeout = 2.0s,
               std::chrono::duration<double> interval = 10ms, const char* message = "condition not met in time") {
    const auto deadline = std::chrono::steady_clock::now() + timeout;
    while (std::chrono::steady_clock::now() < deadline) {
        if (predicate()) return;
        std::this_thread::sleep_for(interval);
    }
    throw std::runtime_error(message);
}

class TempDir {
public:
    TempDir() {
        char tmpl[] = "/tmp/g1bridge_test_XXXXXX";
        const char* dir = mkdtemp(tmpl);
        if (dir == nullptr) throw std::runtime_error("mkdtemp失敗");
        path_ = dir;
    }
    ~TempDir() {
        std::remove((path_ + "/cmd.sock").c_str());
        std::remove((path_ + "/state.sock").c_str());
        ::rmdir(path_.c_str());
    }
    std::string cmd_path() const { return path_ + "/cmd.sock"; }
    std::string state_path() const { return path_ + "/state.sock"; }

private:
    std::string path_;
};

}  // namespace

// 速いテスト条件(cmd_rate=50Hz)で構成する。実運用値は仕様書の20Hzを想定。
class SdkBridgeTest : public ::testing::Test {
protected:
    void SetUp() override {
        cfg_.cmd_sock_path = tmp_.cmd_path();
        cfg_.state_sock_path = tmp_.state_path();
        cfg_.cmd_rate_hz = 50.0;  // period=0.02s
        cfg_.sdk_command_duration_s = 0.04;
        cfg_.cmd_timeout_s = 0.08;
        cfg_.max_sdk_errors = 3;
        bridge_ = std::make_unique<SdkBridgeProcess>(cfg_, backend_);
        bridge_->Start();
    }

    void TearDown() override { bridge_->Stop(); }

    SeqPacketEndpoint ConnectCmdClient() { return ConnectClient(cfg_.cmd_sock_path, sizeof(CmdWire), 1.0); }
    SeqPacketEndpoint ConnectStateClient() { return ConnectClient(cfg_.state_sock_path, sizeof(StateWire), 1.0); }

    void SendCmd(SeqPacketEndpoint& client, std::uint64_t seq, double vx, double vy, double omega) {
        const CmdPacket pkt{seq, MonotonicNs(), vx, vy, omega};
        const CmdWire wire = pkt.Encode();
        client.SendLatest(&wire, sizeof(wire));
    }

    TempDir tmp_;
    SdkBridgeConfig cfg_;
    MockMoveBackend backend_;
    std::unique_ptr<SdkBridgeProcess> bridge_;
};

TEST_F(SdkBridgeTest, StartupSendsZeroBeforeAnyConnection) {
    // D-11: 起動直後、接続前でも必ずゼロ速度を1回送っている
    const auto first = backend_.AllCalls().front();
    EXPECT_DOUBLE_EQ(first.vx, 0.0);
    EXPECT_DOUBLE_EQ(first.vy, 0.0);
    EXPECT_DOUBLE_EQ(first.omega, 0.0);
    EXPECT_DOUBLE_EQ(first.duration, cfg_.sdk_command_duration_s);
}

TEST_F(SdkBridgeTest, DurationArgumentIsAlwaysFinite) {
    // D-27: 呼び出しの度にdurationを明示していることを確認する(continous_move相当を使わない)
    WaitUntil([this]() { return backend_.CallCount() >= 3; });
    for (const auto& call : backend_.AllCalls()) {
        EXPECT_GT(call.duration, 0.0);
        EXPECT_LE(call.duration, cfg_.cmd_timeout_s);
    }
}

TEST_F(SdkBridgeTest, CmdForwardedToBackend) {
    auto client = ConnectCmdClient();
    WaitUntil([this]() { return bridge_->status() != BridgeStatus::kDisconnected; });
    SendCmd(client, 1, 0.15, 0.0, 0.05);

    WaitUntil([this]() {
        const auto last = backend_.LastCall();
        return last.has_value() && last->vx == 0.15 && last->vy == 0.0 && last->omega == 0.05;
    });
}

TEST_F(SdkBridgeTest, WatchdogZeroesAfterCmdTimeout) {
    auto client = ConnectCmdClient();
    WaitUntil([this]() { return bridge_->status() != BridgeStatus::kDisconnected; });
    SendCmd(client, 1, 0.15, 0.0, 0.0);
    WaitUntil([this]() {
        const auto last = backend_.LastCall();
        return last.has_value() && last->vx == 0.15;
    });

    // それ以降は何も送らない -> cmd_timeout超過でSDK側watchdogがゼロを送るはず
    WaitUntil(
        [this]() {
            const auto last = backend_.LastCall();
            return last.has_value() && last->vx == 0.0 && last->vy == 0.0 && last->omega == 0.0;
        },
        1.0s, 10ms, "cmd_timeout超過後もゼロ速度が送られなかった(SDK側watchdogが機能していない)");
}

TEST_F(SdkBridgeTest, PeerDisconnectSetsStatusDisconnectedAndZero) {
    auto client = ConnectCmdClient();
    WaitUntil([this]() { return bridge_->status() != BridgeStatus::kDisconnected; });
    SendCmd(client, 1, 0.1, 0.0, 0.0);
    WaitUntil([this]() {
        const auto last = backend_.LastCall();
        return last.has_value() && last->vx == 0.1;
    });

    client.Close();
    WaitUntil([this]() { return bridge_->status() == BridgeStatus::kDisconnected; });
    WaitUntil([this]() {
        const auto last = backend_.LastCall();
        return last.has_value() && last->vx == 0.0;
    });
}

TEST_F(SdkBridgeTest, ConsecutiveSdkErrorsTriggerFaultAndStayUntilClear) {
    auto client = ConnectCmdClient();
    WaitUntil([this]() { return bridge_->status() != BridgeStatus::kDisconnected; });

    backend_.FailNext(cfg_.max_sdk_errors);
    SendCmd(client, 1, 0.1, 0.0, 0.0);

    WaitUntil([this]() { return bridge_->status() == BridgeStatus::kFault; }, 2.0s, 10ms,
              "連続SDKエラーでFAULTに遷移しなかった");

    // D-13: FAULT中は新しい有効な指令が来ても自動では復帰しない
    SendCmd(client, 2, 0.1, 0.0, 0.0);
    std::this_thread::sleep_for(100ms);
    EXPECT_EQ(bridge_->status(), BridgeStatus::kFault);
    const auto last = backend_.LastCall();
    ASSERT_TRUE(last.has_value());
    EXPECT_DOUBLE_EQ(last->vx, 0.0);

    // 操作者がClearFault()した後は復帰できる
    EXPECT_TRUE(bridge_->ClearFault());
    SendCmd(client, 3, 0.1, 0.0, 0.0);
    WaitUntil([this]() {
        const auto l = backend_.LastCall();
        return l.has_value() && l->vx == 0.1;
    });
}

TEST_F(SdkBridgeTest, SuccessfulCallResetsConsecutiveErrorCount) {
    // 仕様書8章: 「連続」失敗回数。成功が挟まればカウントはリセットされ、FAULTにならない
    auto client = ConnectCmdClient();
    WaitUntil([this]() { return bridge_->status() != BridgeStatus::kDisconnected; });

    for (int i = 0; i < cfg_.max_sdk_errors - 1; ++i) {
        backend_.FailNext(1);
        SendCmd(client, 1, 0.1, 0.0, 0.0);
        WaitUntil([this]() { return bridge_->sdk_error_count() >= 1; });
        WaitUntil([this]() { return bridge_->sdk_error_count() == 0; });  // 次の成功呼び出しでリセットされる
    }

    EXPECT_NE(bridge_->status(), BridgeStatus::kFault);
}

TEST_F(SdkBridgeTest, StateChannelReportsStatusAndPose) {
    auto cmd_client = ConnectCmdClient();
    auto state_client = ConnectStateClient();
    WaitUntil([this]() { return bridge_->status() != BridgeStatus::kDisconnected; });

    SendCmd(cmd_client, 1, 0.1, 0.0, 0.0);

    std::vector<StatePacket> received;
    const auto deadline = std::chrono::steady_clock::now() + 1s;
    while (std::chrono::steady_clock::now() < deadline && received.size() < 3) {
        auto raw = state_client.RecvLatest();
        if (raw.has_value()) {
            received.push_back(StatePacket::Decode(raw->data(), raw->size()));
        }
        std::this_thread::sleep_for(5ms);
    }

    ASSERT_GE(received.size(), 1u);
    EXPECT_TRUE(std::any_of(received.begin(), received.end(),
                             [](const StatePacket& p) { return p.status == BridgeStatus::kNavigating; }));
    // x=0スタートで前進指令を送っているので、いずれかの時点でxが正に進んでいるはず
    EXPECT_TRUE(std::any_of(received.begin(), received.end(), [](const StatePacket& p) { return p.x > 0.0; }));
}

File: Navigation/nav2_option/g1_sdk_bridge/tests/test_sdk_bridge.py
"""SdkBridgeProcess の統合テスト。実スレッド + 実Unix domain socketで検証する。

タイミング系のテストはsleepではなくwait_until()でポーリングし、CI環境差による
flakyさを抑える。ただしスレッド+実ソケットを使うため、他のテストより本質的に
時間がかかる(数百ms程度)。
"""

import os
import tempfile
import time
import unittest

import _pathfix  # noqa: F401
from ipc_transport import connect_client
from protocol import BridgeStatus, CmdPacket, StatePacket, monotonic_ns
from sdk_process_mock import MockMoveBackend, SdkBridgeConfig, SdkBridgeProcess


def wait_until(predicate, timeout=2.0, interval=0.01, message="condition not met in time"):
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if predicate():
            return
        time.sleep(interval)
    raise AssertionError(message)


class SdkBridgeTestCase(unittest.TestCase):
    """速いテスト条件(cmd_rate=50Hz)で構成する。実運用値は仕様書の20Hzを想定。"""

    def setUp(self):
        self._tmpdir = tempfile.mkdtemp(prefix="g1bridge_")
        self._cmd_path = os.path.join(self._tmpdir, "cmd.sock")
        self._state_path = os.path.join(self._tmpdir, "state.sock")
        self._backend = MockMoveBackend()
        self._cfg = SdkBridgeConfig(
            cmd_sock_path=self._cmd_path,
            state_sock_path=self._state_path,
            cmd_rate_hz=50.0,  # period=0.02s
            sdk_command_duration_s=0.04,
            cmd_timeout_s=0.08,
            max_sdk_errors=3,
        )
        self._bridge = SdkBridgeProcess(self._cfg, self._backend)
        self._bridge.start()

    def tearDown(self):
        self._bridge.stop()

    def _connect_cmd_client(self):
        return connect_client(self._cmd_path, max_payload=64, timeout=1.0)

    def _connect_state_client(self):
        return connect_client(self._state_path, max_payload=96, timeout=1.0)

    def _send_cmd(self, client, seq, vx, vy, omega):
        client.send_latest(CmdPacket(seq=seq, timestamp_ns=monotonic_ns(), vx=vx, vy=vy, omega=omega).encode())

    def test_startup_sends_zero_before_any_connection(self):
        # D-11: 起動直後、接続前でも必ずゼロ速度を1回送っている
        first = self._backend.calls[0]
        self.assertEqual(first[:3], (0.0, 0.0, 0.0))
        self.assertEqual(first[3], self._cfg.sdk_command_duration_s)

    def test_duration_argument_is_always_finite(self):
        # D-27: 呼び出しの度にdurationを明示していることを確認する(continous_move相当を使わない)
        wait_until(lambda: len(self._backend.calls) >= 3)
        for call in self._backend.calls:
            duration = call[3]
            self.assertGreater(duration, 0.0)
            self.assertLessEqual(duration, self._cfg.cmd_timeout_s)

    def test_cmd_forwarded_to_backend(self):
        client = self._connect_cmd_client()
        wait_until(lambda: self._bridge.status != BridgeStatus.DISCONNECTED)
        self._send_cmd(client, 1, 0.15, 0.0, 0.05)

        wait_until(lambda: self._backend.last_call()[:3] == (0.15, 0.0, 0.05))
        client.close()

    def test_watchdog_zeroes_after_cmd_timeout(self):
        client = self._connect_cmd_client()
        wait_until(lambda: self._bridge.status != BridgeStatus.DISCONNECTED)
        self._send_cmd(client, 1, 0.15, 0.0, 0.0)
        wait_until(lambda: self._backend.last_call()[:3] == (0.15, 0.0, 0.0))

        # それ以降は何も送らない -> cmd_timeout超過でSDK側watchdogがゼロを送るはず
        wait_until(
            lambda: self._backend.last_call()[:3] == (0.0, 0.0, 0.0),
            timeout=1.0,
            message="cmd_timeout超過後もゼロ速度が送られなかった(SDK側watchdogが機能していない)",
        )
        client.close()

    def test_peer_disconnect_sets_status_disconnected_and_zero(self):
        client = self._connect_cmd_client()
        wait_until(lambda: self._bridge.status != BridgeStatus.DISCONNECTED)
        self._send_cmd(client, 1, 0.1, 0.0, 0.0)
        wait_until(lambda: self._backend.last_call()[:3] == (0.1, 0.0, 0.0))

        client.close()
        wait_until(lambda: self._bridge.status == BridgeStatus.DISCONNECTED)
        wait_until(lambda: self._backend.last_call()[:3] == (0.0, 0.0, 0.0))

    def test_consecutive_sdk_errors_trigger_fault_and_stay_until_clear(self):
        client = self._connect_cmd_client()
        wait_until(lambda: self._bridge.status != BridgeStatus.DISCONNECTED)

        self._backend.fail_next(self._cfg.max_sdk_errors)
        self._send_cmd(client, 1, 0.1, 0.0, 0.0)

        wait_until(
            lambda: self._bridge.status == BridgeStatus.FAULT,
            message="連続SDKエラーでFAULTに遷移しなかった",
        )

        # D-13: FAULT中は新しい有効な指令が来ても自動では復帰しない
        self._send_cmd(client, 2, 0.1, 0.0, 0.0)
        time.sleep(0.1)
        self.assertEqual(self._bridge.status, BridgeStatus.FAULT)
        self.assertEqual(self._backend.last_call()[:3], (0.0, 0.0, 0.0))

        # 操作者がclear_fault()した後は復帰できる
        self.assertTrue(self._bridge.clear_fault())
        self._send_cmd(client, 3, 0.1, 0.0, 0.0)
        wait_until(lambda: self._backend.last_call()[:3] == (0.1, 0.0, 0.0))

        client.close()

    def test_successful_call_resets_consecutive_error_count(self):
        # 仕様書8章: 「連続」失敗回数。成功が挟まればカウントはリセットされ、FAULTにならない
        client = self._connect_cmd_client()
        wait_until(lambda: self._bridge.status != BridgeStatus.DISCONNECTED)

        for _ in range(self._cfg.max_sdk_errors - 1):
            self._backend.fail_next(1)
            self._send_cmd(client, 1, 0.1, 0.0, 0.0)
            wait_until(lambda: self._bridge.sdk_error_count >= 1)
            wait_until(lambda: self._bridge.sdk_error_count == 0)  # 次の成功呼び出しでリセットされる

        self.assertNotEqual(self._bridge.status, BridgeStatus.FAULT)
        client.close()

    def test_state_channel_reports_status_and_pose(self):
        cmd_client = self._connect_cmd_client()
        state_client = self._connect_state_client()
        wait_until(lambda: self._bridge.status != BridgeStatus.DISCONNECTED)

        self._send_cmd(cmd_client, 1, 0.1, 0.0, 0.0)

        received = []
        deadline = time.monotonic() + 1.0
        while time.monotonic() < deadline and len(received) < 3:
            raw = state_client.recv_latest()
            if raw is not None:
                received.append(StatePacket.decode(raw))
            time.sleep(0.005)

        self.assertGreaterEqual(len(received), 1)
        self.assertTrue(any(p.status == BridgeStatus.NAVIGATING for p in received))
        # x=0スタートで前進指令を送っているので、いずれかの時点でxが正に進んでいるはず
        self.assertTrue(any(p.x > 0.0 for p in received))

        cmd_client.close()
        state_client.close()


if __name__ == "__main__":
    unittest.main()
